src/vs-picgo/index.ts

```diff
@@ -0,0 +1,192 @@
+
+import { formatParam, formatString } from '../utils';
+
+const _ = require('lodash');
+const _db = require('lodash-id');
+_.mixin(_db);
+
+const writeFileP = promisify(fs.writeFile);
+const readFileP = promisify(fs.readFile);
+
+export interface INotice {
+  body: string;
+  text: string;
+  title: string;
+}
+
+export interface IUploadName {
+  date: string;
+  dateTime: string;
+  fileName: string;
+  extName: string;
+  mdFileName: string;
+  [key: string]: string;
+}
+
+export interface IOutputUrl {
+  uploadedName: string;
+  url: string;
+  [key: string]: string;
+}
+
+export default class VSPicgo {
+  private dataPath: string;
+  private editor: vscode.TextEditor;
+  private picgo: PicGo;
+
+  constructor(editor: vscode.TextEditor) {
+    const picgoConfigPath = vscode.workspace.getConfiguration('picgo').get<string>('configPath');
+    if (picgoConfigPath) {
+      this.picgo = new PicGo(picgoConfigPath);
+    } else {
+      this.picgo = new PicGo();
+      const picBed = vscode.workspace.getConfiguration('picgo.picBed');
+      this.picgo.setConfig({ picBed });
+    }
+    this.editor = editor;
+    this.dataPath = this.getDataPath();
+    // Before upload, we change names of the images.
+    this.addRenameListener();
+    // After upload, we use the custom output format.
+    this.addGenerateOutputListener();
+  }
+
+  addGenerateOutputListener() {
+    this.picgo.on('finished', async (ctx: PicGo) => {
+      let urlText = '';
+      const outputFormatTemplate =
+        vscode.workspace.getConfiguration('picgo').get<string>('customOutputFormat') || '![${uploadedName}](${url})';
+      try {
+        urlText = ctx.output.reduce((acc: string, cur: ImgInfo): string => {
+          return `${acc}${formatString(outputFormatTemplate, {
+            uploadedName: cur.fileName || '',
+            url: cur.imgUrl,
+          })}\n`;
+        }, '');
+        urlText = urlText.trim();
+        await this.updateLog(ctx.output);
+      } catch (err) {
+        if (err instanceof SyntaxError) {
+          vscode.window.showErrorMessage(
+            `The data file ${this.dataPath} has syntax error, ` +
+              `please fix the error by yourself or delete the data file and vs-picgo will recreate for you.`,
+          );
+        } else {
+          vscode.window.showErrorMessage(`Failed to read from data file ${this.dataPath}: ${err || ''}`);
+        }
+        return;
+      }
+      this.editor.edit(textEditor => {
+        textEditor.replace(this.editor.selection, urlText);
+        vscode.window.showInformationMessage('Upload successfully');
+      });
+    });
+  }
+
+  addRenameListener() {
+    this.picgo.on('beforeUpload', (ctx: PicGo) => {
+      const userDefineName = this.getImageName();
+      const uploadNameTemplate =
+        vscode.workspace.getConfiguration('picgo').get<string>('customUploadName') || '${fileName}';
+      if (ctx.output.length === 1) {
+        ctx.output[0].fileName = this.changeFilename(
```

Comment:
You forget the file's `extname`. For example, if the user changes the original name `yyy.jpg` to `xxx`, the `extname`->`.jpg` will be missed.
You can get the `extname` by `ctx.output[0].extname`.
Docs:
![image](https://user-images.githubusercontent.com/12621342/56467942-1395e100-6458-11e9-803b-6b31382b242a.png)

Reply:
I hesitate about this. Finally, i decide to give the whole control of the name to users, including extname.

Reply:
Then if you select some text as your fileName, the `extname` will often be missed

Reply:
Yep, it is.  Indeed, it is not very nice. My decision about this changes.

Reply:
How about both dividing `fileName` into `basename` and `extname`?

Reply:
The `${selection}` is good for `finished` but bad for `upload`.

Reply:
Yes, we don't use it in uploaded name, but the output.

Reply:
User should use `vs-picgo` without any configuration. Any other configurations are optional.

Reply:
For now, I think a better way is the following:

1. If you choose to upload a `xxx.jpg` with selection `my-photo`
2. Change the picture's file name to `my-photo.jpg` before uploading.
3. Output to vscode editor: `![my photo](https://xxx.com/my photo.jpg)` by default.

Reply:
@Molunerfinn I agree with you, the simpler, the better. Let's assume  people do not want to change the extension name before uploading:
![image](https://user-images.githubusercontent.com/24741764/56830398-cde27980-6898-11e9-9748-7e928d2d95a4.png)

Reply:
1. When there are multiple images selected:
![image](https://user-images.githubusercontent.com/24741764/56831822-c7560100-689c-11e9-9adc-a26514407b7a.png)
![image](https://user-images.githubusercontent.com/24741764/56831836-ce7d0f00-689c-11e9-973f-42cf09dafc04.png)
1. When there are multiple images selected & selection used:
![image](https://user-images.githubusercontent.com/24741764/56831870-ee143780-689c-11e9-890a-f063d8cce940.png)